libraries/chain/peer_keys_db.cpp

```diff
@@ -0,0 +1,74 @@
+#include <eosio/chain/peer_keys_db.hpp>
+#include <eosio/chain/contract_table_objects.hpp>
+
+namespace eosio::chain {
+
+peer_keys_db_t::peer_keys_db_t()
+   : _peer_key_map(boost::make_shared<peer_key_map_t>()) {
+}
+
+std::optional<uint64_t> peer_keys_db_t::_get_version(const chainbase::database& db) {
+   const auto* t_id =
+      db.find<table_id_object, by_code_scope_table>(boost::make_tuple("eosio"_n, "eosio"_n, "peerkeysver"_n));
+   if (t_id == nullptr)
+      return std::optional<uint64_t>{};
+
+   const auto& idx = db.get_index<chain::key_value_index, chain::by_scope_primary>();
+
+   auto itr = idx.lower_bound( boost::make_tuple(t_id->id, 0));
+   if (itr == idx.end() || itr->t_id != t_id->id || itr->primary_key != 0)
+      return std::optional<uint64_t>{};
+   
+   const key_value_object& obj = *itr;
+   fc::datastream<const char*> ds(obj.value.data(), obj.value.size());
+   uint64_t res = 0;
+   fc::raw::unpack(ds, res);
+   return std::optional<uint64_t>(res);
+}
+
+size_t peer_keys_db_t::update_peer_keys(const controller& chain) {
+   size_t num_updated = 0;
+   try {
+      const auto& db = chain.db();
+      auto cb_version = _get_version(db);
+      if (!cb_version || *cb_version == _version)
+         return num_updated;                    // nothing to do
+      assert(*cb_version > _version);
```

Comment:
Won't this potentially assert on fork switches?

Reply:
Hum, good point, let me think about it.

Reply:
Should be addressed by https://github.com/AntelopeIO/spring/pull/1228/commits/2bcfc00b690c2ab997e24c3815486139db0655d8 and 
https://github.com/eosnetworkfoundation/eos-system-contracts/pull/185/commits/90773cb692cbb5f07c83db3924f9ef62ad9ca3fe

which now update the reflected maps only up to lib.

Reply:
Doesn't matter what signal you attach to, the state of the node is as of HEAD unless you are running in IRREVERSIBLE mode.

Reply:
Sorry, I misunderstood your comment. You are only pulling rows with block numbers up to lib from the table.